### ISSUE TITLE
Error: Invalid resource instance data in state

### ISSUE THREAD

**Issue by heydonovan** [hashicorp/terraform#22699]
I'm still working on slowly upgrading our terraform repo from 0.11 to 0.12. The only part we continue to struggle with are data sources. Guessing because if two folders are using the same data source, one will bust once we upgrade to 0.12, right? Documentation does state that upgrading to 0.11.14 should help this issue.

We are using the rds module, which has been upgraded from `~> 1.0` to `~> 2.0`. Upon running `terraform 0.12upgrade -yes`, the state file doesn't change to reflect this new version (guessing only after a `terraform apply`). A `terraform plan` gives this:

```
Error: Invalid value for module argument

  on main.tf line 42, in module "db":
  42:   subnet_ids = [data.terraform_remote_state.vpc.outputs.database_subnets]

The given value is not suitable for child module variable "subnet_ids" defined
at
.terraform/modules/db/terraform-aws-modules-terraform-aws-rds-454dba6/variables.tf:226,1-22:
element 0: string required.

Releasing state lock. This may take a few moments...
```

I'm trying to see this value by running `terraform console`:

```
> data.terraform_remote_state.vpc.outputs.database_subnets

>
Error: Invalid resource instance data in state

  on /path/to/dir/data.tf line 1:
   1: data "terraform_remote_state" "vpc" {

Instance data.terraform_remote_state.vpc data could not be decoded from the
state: cannot decode dynamic from flatmap.
```

I'll post the `terraform.tfstate` if need be, just not sure what info to strip from it.

Edit: This is not limited to this one folder. I'm running into this issue quite often, so I'm wondering if I'm going about this upgrade process all wrong.  If I had to guess, I'd have to bet it's based on the fact that in 0.11 you could have resources that were referred to by a number, using the count variable. Wonder if 0.12 is expecting each one to have its own name and resource block, so it can actually reference them. Cause if I go into the folder where the data source is pulled from, a `terraform output` works just fine, but the state file is somehow busted when other resources are calling it. :/

Another example:

```
❯ terraform console
> data.terraform_remote_state.nat_gateway.outputs.nat_public_ips


> Error: Invalid resource instance data in state

  on /Users/donovanhernandez/projects/terraform/vpc_madness/vpc/data.tf line 1:
   1: data "terraform_remote_state" "nat_gateway" {

Instance data.terraform_remote_state.nat_gateway data could not be decoded
from the state: cannot decode dynamic from flatmap.
```

Here is the state file: https://gist.github.com/heydonovan/8c03954ba36855c3da9ad7c5ff3e72ec

**Comment by hashibot** [comment]
Hello! :wave: 

This issue doesn't seem to be following one of the standard issue templates available from [the "New Issue" page](https://github.com/hashicorp/terraform/issues/new/choose). The information requested in those templates is important for my human friends to fully understand what's going on.

Please feel free to open a new issue using either the Bug Report or Feature Request template, as appropriate, and then the Terraform maintainers will review it. Thanks!

Please note that we use GitHub issues for tracking bugs and enhancements rather than for questions. While we may be able to help with certain simple problems here it's better to use [the community forum](https://discuss.hashicorp.com/c/terraform-core), where there are more people ready to help. The GitHub issues here are generally monitored only by our few core maintainers.
